spaceship/lib/spaceship/tunes/iap_detail.rb

```diff
@@ -0,0 +1,192 @@
+        def factory(attrs)
+          return self.new(attrs)
+        end
+      end
+
+      # Saves the current In-App-Purchase
+      def save!
+        # Transform localization versions back to original format.
+        versions_array = []
+        versions.each do |language, value|
+          versions_array << {
+                    value: {
+                      description: { value: value[:description] },
+                      name: { value: value[:name] },
+                      localeCode: language.to_s
+                    }
+          }
+        end
+        raw_data["versions"][0]["details"]["value"] = versions_array
+        raw_data['versions'][0]["reviewNotes"] = { value: @review_notes } if @review_notes
+
+        # transform pricingDetails
+        intervals_array = []
+        pricing_intervals.each do |interval|
+          intervals_array << {
+            value: {
+              tierStem: interval[:tier],
+              priceTierEffectiveDate: interval[:begin_date],
+              priceTierEndDate: interval[:end_date],
+              country: interval[:country] || "WW",
+              grandfathered: interval[:grandfathered]
+            }
+          }
+        end
+
+        if subscription_price_target
+          intervals_array = []
+          pricing_calculator = client.iap_subscription_pricing_target(app_id: application.apple_id, purchase_id: purchase_id, currency: subscription_price_target[:currency], tier: subscription_price_target[:tier])
+          pricing_calculator.each do |language_code, value|
+            intervals_array << {
+              value: {
+                tierStem: value["tierStem"],
+                priceTierEffectiveDate: value["priceTierEffectiveDate"],
+                priceTierEndDate: value["priceTierEndDate"],
+                country: language_code,
+                grandfathered: { value: "FUTURE_NONE" }
+              }
+            }
+          end
+
+        end
+
+        raw_data.set(["pricingIntervals"], intervals_array)
+
+        if @review_screenshot
+          # Upload Screenshot
+          upload_file = UploadFile.from_path @review_screenshot
+          screenshot_data = client.upload_purchase_review_screenshot(application.apple_id, upload_file)
+          new_screenshot = {
+            "value" => {
+              "assetToken" => screenshot_data["token"],
+              "sortOrder" => 0,
+              "type" => "SortedScreenShot",
+              "originalFileName" => upload_file.file_name,
+              "size" => screenshot_data["length"],
+              "height" => screenshot_data["height"],
+              "width" => screenshot_data["width"],
+              "checksum" => screenshot_data["md5"]
+            }
+          }
+
+          raw_data["versions"][0]["reviewScreenshot"] = new_screenshot
+        end
+        # Update the Purchase
+        client.update_iap!(app_id: application.apple_id, purchase_id: self.purchase_id, data: raw_data)
+      end
+
+      def setup
```

Comment:
first of all - thx for taking the time to review this.

i need to get and transform values from the response that are too deep nested - to handle them with `attr_mapping` (atleast i didn't find a way to do it with that, as i would have loved to do it with mapped attributes.)

i was looking through the code and found a couple of places where similar problems where solved via `setup` (`application.rb`, `app_version.rb`, `tester.rb`.....) - and on my tests it seemed to work too. so i did it that way.

if there is a better way todo it, not sure on how to do it in factory (or is it just a code move?) - it would be super awesome if you could point me to samples doing stuff like this in the factory.